File: Week6_Strings/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        # Complement map to save space
        comp_map = {"(" : ")", "[" : "]", "{" : "}"}
        # Unnecessary actually. Can use keys of comp_map
        open_set = {"(", "[", "{"}
        # Traverse string from left to right
        for char in s:
            if char in open_set:
                # If character classified as left bracket,
                # Push it into the stack
                stack.append(char)
            # This branch focuses on right brackets
            else:
                # No available matching left brackets
                if len(stack) == 0:
                    return False
                # If there is left bracket in stack,
                # Check if it's a suitable complement
                comp = stack.pop()
                if comp_map[comp] != char:
                    return False
        
        return len(stack) == 0

File: Week6_Strings/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        /* Complement map to map bracket to its conjugate */
        Map<Character, Character> complement = new HashMap<>();
        complement.put('(', ')');
        complement.put('[', ']');
        complement.put('{', '}');
        
        /* Initialize stack */
        Stack<Character> stk = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            /**
             * If character classified as left bracket
             * push it into the stack
             */ 
            if (complement.containsKey(ch)) {
                stk.add(ch);
            }
            /* In this branch, ch is a right bracket */
            else {
                /* No available matching left brackets */
                if (stk.isEmpty()) return false;
                char comp = stk.pop();
                /**
                 * Check if left bracket popped from stack
                 * is indeed a suitable complement
                 */
                if (complement.get(comp) != ch) return false;
            }
        }
        
        /* No stray left brackets */
        return stk.isEmpty();
    }
}